packages/graphql-codegen-cli/tests/watcher.spec.ts

```diff
@@ -0,0 +1,142 @@
+      await setupMockWatcher({
+        filepath: './foo/some-config.ts',
+        config: {
+          schema: './foo/something.ts',
+          generates: {
+            ['./foo/some-output.ts']: {
+              documents: ['./foo/bar/*.graphql'],
+            },
+          },
+        },
+      });
+
+    expect(watchDirectory).toBe(join(process.cwd(), 'foo'));
+
+    const shouldTriggerBuild = join(process.cwd(), './foo/bar/fizzbuzz.graphql');
+    const shouldNotTriggerBuild = join(process.cwd(), './foo/bar/something.ts');
+
+    await dispatchChange(shouldTriggerBuild);
+    expect(subscribeCallbackSpy).toHaveBeenLastCalledWith(undefined, [{ path: shouldTriggerBuild, type: 'update' }]);
+    expect(onWatchTriggered).toHaveBeenLastCalledWith('update', shouldTriggerBuild);
+
+    await dispatchChange(shouldNotTriggerBuild);
+    expect(subscribeCallbackSpy).toHaveBeenLastCalledWith(undefined, [{ path: shouldNotTriggerBuild, type: 'update' }]);
+    expect(onWatchTriggered).not.toHaveBeenLastCalledWith('update', shouldNotTriggerBuild);
+    expect(onWatchTriggered).toHaveBeenCalledTimes(1);
+
+    expect(subscribeCallbackSpy).toHaveBeenCalledTimes(2);
+
+    await stopWatching();
+    expect(unsubscribeSpy).toHaveBeenCalled();
+  });
+
+  test('triggers rebuilds as expected (auto-assertions)', async () => {
+    const mockWatcher = await setupMockWatcher({
+      filepath: './foo/some-config.ts',
+      config: {
+        schema: './foo/something.ts',
+        generates: {
+          ['./foo/some-output.ts']: {
+            documents: ['./foo/bar/*.graphql'],
+          },
+          ['./foo/some-other-output.ts']: {
+            documents: ['./foo/some-other-bar/*.graphql'],
+          },
+          ['./foo/some-preset-bar/']: {
+            preset: 'near-operation-file',
+            presetConfig: {
+              extension: '.generated.tsx',
+              baseTypesPath: 'types.ts',
+            },
+            documents: ['./foo/some-preset-bar/*.graphql'],
+          },
+        },
+      },
+    });
+
+    await assertBuildTriggers(mockWatcher, {
+      shouldTriggerBuild: [
+        './foo/some-config.ts', // config file
+        './foo/bar/fizzbuzz.graphql',
+      ],
+      pathsWouldBeIgnoredByParcelWatcher: [
+        // note: expectations should be relative to cwd; assertion helper converts
+        //       the values received by parcelWatcher to match before testing them (see typedoc)
+        './foo/some-output.ts', // output file
+        'foo/some-output.ts', // output file
+      ],
+      globsWouldBeIgnoredByParcelWatcher: [
+        // note: globs are tested for exact match with argument passed to subscribe options
+        'some-preset-bar/**/*.generated.tsx', // output of preset
+      ],
+      shouldNotTriggerBuild: [
+        './foo/bar/something.ts', // unrelated file
+        './foo/some-output.ts', // output file (note: should be ignored by parcel anyway)
+        '.git/index.lock',
+      ],
+    });
+  });
```

Comment:
Ok, so I got.... a bit carried away building the test infrastructure. 👀 I made a few minor changes to the watcher itself, but mostly just for supporting the test infrastructure. At this point the bulk of the PR is all test code.

The watch server is now fully mocked and testable. It can also be cleanly cancelled without any unhandled promises. And I added a whole bunch of helpers for making assertions on whether or not a file change event triggers a rebuild for a given config. That means we can write code like I've highlighted with this comment, which makes for really clean testing.

I think I'm now mostly done with the test infrastructure, and now I just need to add a bunch of cases and actually test it. 😅 But I've found and fixed a few bugs so far, so it's a good start.

Reply:
Here are some examples of what the error message formatting looks like (click to expand for screenshots):

<details>
<summary>Expected file to trigger build, but it didn't</summary>

![image](https://user-images.githubusercontent.com/835921/230686541-65514eac-ea53-4696-b49b-c10024bec8f7.png)

</details>


<details>
<summary>Expected file NOT to trigger build, but it did</summary>

![image](https://user-images.githubusercontent.com/835921/230686701-036e1b78-dfa4-4836-917e-4bf1148b9742.png)


</details>


<details>
<summary>Expected path to be passed to ParcelWatcher.Options.ignore, but it was not</summary>

See: [ParcelWatcher `ignore` option](https://github.com/parcel-bundler/watcher#options)


![image](https://user-images.githubusercontent.com/835921/230686825-c577feee-3eaf-4a70-a566-83de79546a43.png)


</details>


<details>
<summary>Expected glob to be passed to ParcelWatcher.Options.ignore, but it was not</summary>

See: [ParcelWatcher `ignore` option](https://github.com/parcel-bundler/watcher#options)

![image](https://user-images.githubusercontent.com/835921/230686890-f095a62d-0d01-4b11-8822-be3737da331d.png)


</details>